src/relay/FullRelay.sol

```diff
@@ -0,0 +1,461 @@
+    /// @return         The height of the header, or error if unknown
+    function findHeight(bytes32 _digest) external view returns (uint256) {
+        return _findHeight(_digest);
+    }
+
+    /// @notice         Finds an ancestor for a block by its digest
+    /// @dev            Will fail if the header is unknown
+    /// @param _digest  The header digest to search for
+    /// @return         The height of the header, or error if unknown
+    function findAncestor(bytes32 _digest, uint256 _offset) external view returns (bytes32) {
+        return _findAncestor(_digest, _offset);
+    }
+
+    /// @notice             Checks if a digest is an ancestor of the current one
+    /// @dev                Limit the amount of lookups (and thus gas usage) with _limit
+    /// @param _ancestor    The prospective ancestor
+    /// @param _descendant  The descendant to check
+    /// @param _limit       The maximum number of blocks to check
+    /// @return             true if ancestor is at most limit blocks lower than descendant, otherwise false
+    function isAncestor(bytes32 _ancestor, bytes32 _descendant, uint256 _limit) external view returns (bool) {
+        return _isAncestor(_ancestor, _descendant, _limit);
+    }
+
+    /// @notice             Adds headers to storage after validating
+    /// @dev                We check integrity and consistency of the header chain
+    /// @param  _anchor     The header immediately preceeding the new chain
+    /// @param  _headers    A tightly-packed list of 80-byte Bitcoin headers
+    /// @return             True if successfully written, error otherwise
+    function addHeaders(bytes calldata _anchor, bytes calldata _headers) external returns (bool) {
+        require(isHeaderChainValidLength(_headers), "Header array length must be divisible by 80");
+        require(isHeaderValidLength(_anchor), "Anchor must be 80 bytes");
+
+        return _addHeaders(_anchor, _headers, false);
+    }
+
+    /// @notice                       Adds headers to storage, performs additional validation of retarget
+    /// @dev                          Checks the retarget, the heights, and the linkage
+    /// @param  _oldPeriodStartHeader The first header in the difficulty period being closed
+    /// @param  _oldPeriodEndHeader   The last header in the difficulty period being closed
+    /// @param  _headers              A tightly-packed list of 80-byte Bitcoin headers
+    /// @return                       True if successfully written, error otherwise
+    function addHeadersWithRetarget(
+        bytes calldata _oldPeriodStartHeader,
+        bytes calldata _oldPeriodEndHeader,
+        bytes calldata _headers
+    ) external returns (bool) {
+        require(
+            isHeaderValidLength(_oldPeriodStartHeader) && isHeaderValidLength(_oldPeriodEndHeader)
+                && isHeaderChainValidLength(_headers),
+            "Bad args. Check header and array byte lengths."
```

Comment:
For the errors I think you suggested before @ferencdg about using some custom error type? Do we want to add that for this PR or keep the diff with upstream minimal?